src/sokoban.hpp

```diff
@@ -53,14 +53,15 @@ class Sokoban {
     unsigned int py;
     unsigned int px;
     std::vector<Direction> moves;
-    std::vector<std::vector<std::string>> history;
+    std::vector<std::pair<std::vector<std::string>, bool>> history;
     std::vector<std::pair<Direction, std::vector<std::string>>> undone;
 
     void locate_player();
     void move_player(int dy, int dx);
     void push_box(int dy, int dx);
-    void update(Direction direction);
-    bool make_move(Direction direction);
+    void update(Direction direction, bool fast_forward);
```

Comment:
This appears to be a code smell identified by Martin Fowler. See https://martinfowler.com/bliki/FlagArgument.html and [this SE thread](https://softwareengineering.stackexchange.com/questions/147977/is-it-wrong-to-use-a-boolean-parameter-to-determine-behavior) for details. One reason to avoid it is for the caller, in an interface like `foo.bar(something, false)` it's almost impossible to tell what "false" means without consulting the documentation for the API.

I'd probably just remove the flag and make all behavior work like rewind, unless there's an easy way to keep both and refactor them into separate functions `rewind` and `undo` in a way that doesn't involve too much repeated logic, or buries the flag as an implementation detail behind the public API.